File: objectpath-java/src/main/java/cz/zakjan/objectpath/visitor/GetByPathVisitorFunctions.java
package cz.zakjan.objectpath.visitor;

import java.time.Instant;
import java.time.OffsetDateTime;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.function.Function;
import java.util.stream.Collectors;

public class GetByPathVisitorFunctions {
    private static Map<String, Function<List<Object>, Object>> nameToFunction = new HashMap<String, Function<List<Object>, Object>>() {{
        put("toString", GetByPathVisitorFunctions::toStringFunction);
        put("toNumber", GetByPathVisitorFunctions::toNumberFunction);
        put("join", GetByPathVisitorFunctions::joinFunction);
        put("split", GetByPathVisitorFunctions::splitFunction);
        put("sum", GetByPathVisitorFunctions::sumFunction);
        put("dateTimestampToIsoString", GetByPathVisitorFunctions::dateTimestampToIsoStringFunction);
        put("dateIsoStringToTimestamp", GetByPathVisitorFunctions::dateIsoStringToTimestampFunction);
    }};

    public static Object callFunction(String name, List<Object> args) {
        Function<List<Object>, Object> func = nameToFunction.get(name);
        if (func == null) {
            return null;
        }

        return func.apply(args);
    }

    public static Object toStringFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object value = args.get(0);
        String result = String.valueOf(value);
        return result;
    }

    private static Object toNumberFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object value = args.get(0);
        Object result;
        try {
            result = Integer.parseInt(String.valueOf(value));
        } catch (NumberFormatException e) {
            result = Long.parseLong(String.valueOf(value));
        }
        return result;
    }

    private static Object joinFunction(List<Object> args) {
        if (args.size() != 2) {
            return null;
        }

        Object values = args.get(0);
        Object delimiter = args.get(1);
        if (!(values instanceof List && delimiter instanceof String)) {
            return null;
        }

        List<String> filteredValues = ((List<Object>)values).stream().filter(x -> x instanceof String).map(x -> (String)x).collect(Collectors.toList());
        if (filteredValues.size() == 0) {
            return null;
        }

        String result = String.join((String)delimiter, filteredValues);
        return result;
    }

    private static Object splitFunction(List<Object> args) {
        if (args.size() != 2) {
            return null;
        }

        Object value = args.get(0);
        Object delimiter = args.get(1);
        if (!(value instanceof String && delimiter instanceof String)) {
            return null;
        }

        List<String> result = Arrays.asList(((String)value).split((String)delimiter));
        return result;
    }

    private static Object sumFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object values = args.get(0);
        if (!(values instanceof List)) {
            return null;
        }

        List<Integer> filteredValues = ((List<Object>)values).stream().filter(x -> x instanceof Integer).map(x -> (Integer)x).collect(Collectors.toList());
        if (filteredValues.size() == 0) {
            return null;
        }

        Integer result = filteredValues.stream().reduce(0, (sum, x) -> sum + x);
        return result;
    }

    private static Object dateTimestampToIsoStringFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object value = args.get(0);
        if (!(value instanceof Long)) {
            return null;
        }

        OffsetDateTime dateTime = OffsetDateTime.ofInstant(Instant.ofEpochMilli((Long)value), ZoneOffset.UTC);
        return dateTime.format(DateTimeFormatter.ofPattern("YYYY-MM-DD'T'HH:mm:ss.SSSX"));
    }

    private static Object dateIsoStringToTimestampFunction(List<Object> args) {
        if (args.size() != 1) {
            return null;
        }

        Object value = args.get(0);
        if (!(value instanceof String)) {
            return null;
        }

        try {
            OffsetDateTime dateTime = OffsetDateTime.parse((String)value, DateTimeFormatter.ISO_OFFSET_DATE_TIME);
            return dateTime.toInstant().toEpochMilli();
        } catch (DateTimeParseException e) {
            return null;
        }
    }
}

File: objectpath-ts/src/visitor/GetByPathVisitorFunctions.ts
export class GetByPathVisitorFunctions {
    private static nameToFunction = new Map([
        ["toString", GetByPathVisitorFunctions.toStringFunction],
        ["toNumber", GetByPathVisitorFunctions.toNumberFunction],
        ["join", GetByPathVisitorFunctions.joinFunction],
        ["split", GetByPathVisitorFunctions.splitFunction],
        ["sum", GetByPathVisitorFunctions.sumFunction],
        ["dateTimestampToIsoString", GetByPathVisitorFunctions.dateTimestampToIsoStringFunction],
        ["dateIsoStringToTimestamp", GetByPathVisitorFunctions.dateIsoStringToTimestampFunction],
    ]);

    public static callFunction(name: string, args: unknown[]): unknown {
        const func = GetByPathVisitorFunctions.nameToFunction.get(name);
        /* istanbul ignore if */
        if (!func) {
            return null;
        }

        return func(args);
    }

    private static toStringFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        const result = `${value}`;
        return result;
    }

    private static toNumberFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        const result = parseInt(`${value}`, 10);
        return result;
    }

    private static joinFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 2) {
            return null;
        }

        const values = args[0];
        const delimiter = args[1];
        /* istanbul ignore if */
        if (!(Array.isArray(values) && typeof delimiter === "string")) {
            return null;
        }

        const filteredValues = values.filter(x => typeof x === "string");
        /* istanbul ignore if */
        if (filteredValues.length === 0) {
            return null;
        }

        const result = filteredValues.join(delimiter)
        return result;
    }

    private static splitFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 2) {
            return null;
        }

        const value = args[0];
        const delimiter = args[1];
        /* istanbul ignore if */
        if (!(typeof value === "string" && typeof delimiter === "string")) {
            return null;
        }

        const result = value.split(delimiter);
        return result;
    }

    private static sumFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const values = args[0];
        /* istanbul ignore if */
        if (!Array.isArray(values)) {
            return null;
        }

        const filteredValues = values.filter(x => typeof x === "number");
        /* istanbul ignore if */
        if (filteredValues.length === 0) {
            return null;
        }

        const result = filteredValues.reduce((sum, x) => sum + x, 0);
        return result;
    }

    private static dateTimestampToIsoStringFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        /* istanbul ignore if */
        if (typeof value !== "number") {
            return null;
        }

        return new Date(value).toISOString();
    }

    private static dateIsoStringToTimestampFunction(args: unknown[]): unknown {
        /* istanbul ignore if */
        if (args.length !== 1) {
            return null;
        }

        const value = args[0];
        /* istanbul ignore if */
        if (typeof value !== "string") {
            return null;
        }

        return Date.parse(value).valueOf();
    }
}